oximeter/producer/src/lib.rs

```diff
@@ -481,60 +487,89 @@ mod tests {
+            });
+            let task = tokio::spawn(async move {
+                server.serve(make_service).await.unwrap();
+            });
+            Self { task, addr }
+        }
+    }
+
     #[tokio::test]
     async fn test_producer_registration_task() {
         let log = test_logger();
-        let fake_nexus = TestServer::run();
+        let fake_nexus = FakeNexus::new().await;
         slog::info!(
             log,
             "fake nexus test server listening";
-            "address" => ?fake_nexus.addr(),
+            "address" => ?fake_nexus.addr,
         );
 
-        const INTERVAL: Duration = Duration::from_secs(1);
-        let body = ProducerRegistrationResponse { lease_duration: INTERVAL };
-        fake_nexus.expect(
-            Expectation::matching(request::method_path(
-                "POST",
-                "/metrics/producers",
-            ))
-            .times(2..)
-            .respond_with(
-                status_code(201).body(serde_json::to_string(&body).unwrap()),
-            ),
-        );
         let address = "[::1]:0".parse().unwrap();
         let config = Config {
             server_info: ProducerEndpoint {
                 id: Uuid::new_v4(),
                 kind: ProducerKind::Service,
                 address,
-                base_route: String::from("/collect"),
+                base_route: String::new(),
                 interval: Duration::from_secs(10),
             },
-            registration_address: Some(fake_nexus.addr()),
+            registration_address: Some(fake_nexus.addr),
             request_body_max_bytes: 1024,
             log: LogConfig::Logger(log),
         };
 
         // Ideally, we would check pretty carefully that there are exactly N
         // registrations after N renewal periods. That's brittle, especially on
-        // a loaded system. Instead, we'll run for a number of intervals, and
-        // just check that we have more than one.
+        // a loaded system. Instead, we'll wait until we've received the
+        // expected number of registration requests.
         let _server = Server::start(&config).unwrap();
-        let now = Instant::now();
-        tokio::time::pause();
-        let max = INTERVAL * 10;
-        while now.elapsed() < max {
-            tokio::time::advance(Duration::from_millis(10)).await;
-        }
-        tokio::time::resume();
-
-        // Drop the server to check its requests, rather than calling
-        // `verify_and_clear()`. It's possible for requests to be in-flight
-        // between when we called `verify_and_clear()` and when we exit this
-        // test, at which point the drop impl of the `TestServer` checks the
-        // expectations _again_. Let's just do the check by dropping here.
-        drop(fake_nexus);
+        const N_REQUESTS: u32 = 10;
+        const POLL_INTERVAL: Duration = Duration::from_millis(100);
+        const POLL_DURATION: Duration = Duration::from_secs(10);
```

Comment:
Do we maybe need this to be longer? I'm not sure what was causing the old test to falke, but it was also 10x the reregistration interval, right?

Reply:
Sure, we can make it longer. The old test flaked because we were basically racing with the registration task. Here, we're just waiting until we get at least the expected number, so the mechanism is quite different.